src/async_impl/client.rs

```diff
@@ -1239,6 +1247,24 @@ impl ClientBuilder {
         self.config.dns_overrides.insert(domain.to_string(), addr);
         self
     }
+
+    /// Override DNS resolution for specific domains to particular IP addresses.
+    /// If both are specified, overrides has priority.
+    /// Will not be called every time, because results will be cached (if that is correct).
+    ///
+    /// Warning
+    ///
+    /// Since the DNS protocol has no notion of ports, if you wish to send
+    /// traffic to a particular port you must include this port in the URL
+    /// itself, any port in the overridden addr will be ignored and traffic sent
+    /// to the conventional port for the given scheme (e.g. 80 for http).
+    pub fn resolve_fn(
+        mut self,
+        func: Box<dyn Fn(String) -> Option<SocketAddr> + Send + Sync>,
+    ) -> ClientBuilder {
+        self.config.dns_overrides_fn = Some(func);
+        self
+    }
```

Comment:
For ease of use, I think it would be better if this was a generic instead of a boxed `Fn`. The boxing can happen inside this function:

```suggestion
    pub fn resolve_fn(
        mut self,
        func: impl Fn(String) -> Option<SocketAddr> + Send + Sync,
    ) -> ClientBuilder {
        self.config.dns_overrides_fn = Some(Box::new(func));
        self
    }
```

(not tested but something like that should work)

Reply:
I also wonder whether it should be a closure that returns a future, but that depends on the (possible) use cases.

Reply:
> I also wonder whether it should be a closure that returns a future, but that depends on the (possible) use cases.

The call function not is async, seems to have some difficulties return a Future.

```rust
impl Struct for DnsResolverWithOverrides {
     fn call (&mut self, name: Name) -> Self::Future {}
}
```

If the current code, async operations such as downloading IP sets from the network  or read IP sets from database, need to be done before build client.